### PR TITLE
Support adapters between Collections returned by Kotlin and jre_emul collections

### PR DESCRIPTION
**Pull request by guillaumeaudet**
Some methods in kotlin return NSArray for example when accessing a Kotlin List. This is the addition of NSArrayToJavaUtilListAdapter to J2ObjC jre_emul. So we can use them as javaUtilList. This also contains and adapter for the Iterator.

At this moment we only support unmodifiableList so all type of list will crash if you try to modify them.

### CODE REVIEW

File: jre_emul/Classes/NSArrayToJavaUtilListAdapter.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
+
```
Comment (flambert) [inline jre_emul/Classes/NSArrayToJavaUtilListAdapter.h:1]: Why the empty first line?

File: jre_emul/Classes/NSArrayToJavaUtilListAdapter.m [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    return toJavaUtilListIterator(_sourceArray);
+}
+
+- (IOSObjectArray *)toArray {
+    unsupportedAdapterCallWithName(@"toArray");
+    return [IOSObjectArray arrayWithLength:0 type:NSObject_class_()];
+}
+
+- (jboolean)addWithId:(id)o {
+    illegalAdapterMutableCallWithName(@"addWithId");
```
Comment (flambert) [inline jre_emul/Classes/NSArrayToJavaUtilListAdapter.m:40]: `NSStringFromSelector(_cmd)` is not OK?

File: jre_emul/Classes/NSArrayToJavaUtilListAdapter.m [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+- (void)forEachWithJavaUtilFunctionConsumer:(id<JavaUtilFunctionConsumer>)arg0 {
+    JavaLangIterable_forEachWithJavaUtilFunctionConsumer_(self, arg0);
+}
+
+- (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state objects:(__unsafe_unretained id *)stackbuf count:(NSUInteger)len {
+    return JreDefaultFastEnumeration(self, state, stackbuf);
+}
+
+
+
```
Comment (flambert) [inline jre_emul/Classes/NSArrayToJavaUtilListAdapter.m:162]: Why the 3 blank lines?

File: jre_emul/Classes/NSArrayToJavaUtilListAdapter.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+- (id<JavaUtilList> __nonnull)subListWithInt:(jint)i
+                                     withInt:(jint)i1;
+
+- (IOSObjectArray * __nonnull)toArray;
+
+- (IOSObjectArray * __nonnull)toArrayWithNSObjectArray:(IOSObjectArray * __nonnull)objects;
+
+@end
+
+
```
Comment (flambert) [inline jre_emul/Classes/NSArrayToJavaUtilListAdapter.h:90]: Why the 2 blank lines?

File: translator/src/main/java/com/google/devtools/j2objc/translate/KotlinCollectionsConverter.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    public void endVisit(Assignment node) {
+        super.endVisit(node);
+    }
+
+    @Override
+    public boolean visit(ExpressionStatement node) {
+        return super.visit(node);
+    }
+        super.endVisit(node);
+    }
```
Comment (flambert) [inline translator/src/main/java/com/google/devtools/j2objc/translate/KotlinCollectionsConverter.java:51]: Doesn't seem right

File: translator/src/main/java/com/google/devtools/j2objc/translate/KotlinCollectionsConverter.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public boolean visit(Assignment node) {
+        return super.visit(node);
+    }
+
+    @Override
+    public void endVisit(Assignment node) {
+        super.endVisit(node);
+    }
```
Comment (flambert) [inline translator/src/main/java/com/google/devtools/j2objc/translate/KotlinCollectionsConverter.java:44]: Don't override functions for nothing

File: translator/src/main/java/com/google/devtools/j2objc/util/KotlinUtil.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+
+import kotlin.Metadata;
+import kotlinx.metadata.KmClass;
+import kotlinx.metadata.jvm.KotlinClassHeader;
+import kotlinx.metadata.jvm.KotlinClassMetadata;
+
+public final class KotlinUtil {
```
Comment (flambert) [inline translator/src/main/java/com/google/devtools/j2objc/util/KotlinUtil.java:16]: Don't forget the private constructor

File: translator/src/main/java/com/google/devtools/j2objc/util/KotlinUtil.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    public enum KotlinCollectionType {
+        LIST,
+        NONE,
```
Comment (flambert) [inline translator/src/main/java/com/google/devtools/j2objc/util/KotlinUtil.java:20]: NONE should always be first

File: jre_emul/Classes/NSArrayToJavaUtilListAdapter.m (hunk 2 above)
Comment (guillaumeaudet) [reply jre_emul/Classes/NSArrayToJavaUtilListAdapter.m:40]: Did not know this exists. I'll try it !

Comment (guillaumeaudet) [reply jre_emul/Classes/NSArrayToJavaUtilListAdapter.m]: Works like a charm !

File: translator/src/main/java/com/google/devtools/j2objc/translate/KotlinCollectionsConverter.java (hunk 5 above)
Comment (guillaumeaudet) [reply translator/src/main/java/com/google/devtools/j2objc/translate/KotlinCollectionsConverter.java]: indeed, i messed up !

**Review by mbaron — APPROVED** [review]

**Pull request merged**